### PR TITLE
Dockerfiles: bump eve-alpine image, eve-cross-compilers and eve-dom0-ztools

### PR DESCRIPTION
**Pull request by christoph-zededa**
Sync eve-alpine image across all packages with the latest image.

```
> docker inspect lfedge/eve-alpine:1f7685f95a475c6bbe682f0b976f12180b6c8726 | grep -i create
        "Created": "2024-05-14T20:55:51.41185787Z",

> docker inspect lfedge/eve-alpine:82df60e43ab9f8c935584b8c7b4d0a4b0271d608 | grep -i create
        "Created": "2024-09-05T21:14:17.854134914Z",
```

### CODE REVIEW

**Review by OhmSpectator — APPROVED** [review]
We have to automate the check on proper image hashes...

**Comment by deitch** [comment]
> @deitch so after https://github.com/lf-edge/eve/pull/4220#issuecomment-2334242270 now bumping the installer makes problems? ;-)

That dependency from mkverification must have moved into installer. We should check the installer to see if it is something we still actually need.

**Comment by christoph-zededa** [comment]
> We have to automate the check on proper image hashes...

there is https://github.com/lf-edge/eve-build-tools/blob/main/src/dockerfile-from-checker/main.go but it needs some changes

**Comment by OhmSpectator** [comment]
> > We have to automate the check on proper image hashes...
> 
> there is https://github.com/lf-edge/eve-build-tools/blob/main/src/dockerfile-from-checker/main.go but it needs some changes

I though about a GH action that can report inconsistency in the hashes in different components.

**Comment by christoph-zededa** [comment]
> > > We have to automate the check on proper image hashes...
> > 
> > 
> > there is https://github.com/lf-edge/eve-build-tools/blob/main/src/dockerfile-from-checker/main.go but it needs some changes
> 
> I thought about a GH action that can report inconsistency in the hashes in different components.

if you can run it from a GH action, then it should work, just run it with:
```
dockerfile-checker ./ -i ./eve-tools/bpftrace-compiler/Dockerfile -i ./pkg/optee-os/Dockerfile -i pkg/bsp-imx/Dockerfile -i pkg/vtpm/Dockerfile
```

because there are still some deviations:
```
tags differ for image lfedge/eve-dom0-ztools:b8eaeec19d373228a4a842374e5de0d50f050853 in files /home/christoph/projects/eve/pkg/vtpm/Dockerfile and /home/christoph/projects/eve/pkg/pillar/Dockerfile
tags differ for image lfedge/eve-cross-compilers:2a1d062fce410865e7024a83de327a68e52db26c in files /home/christoph/projects/eve/pkg/pillar/Dockerfile and /home/christoph/projects/eve/pkg/optee-os/Dockerfile
tags differ for image alpine:3.20 in files /home/christoph/projects/eve/eve-tools/bpftrace-compiler/Dockerfile and /home/christoph/projects/eve/eve-tools/bpftrace-compiler/root/Dockerfile
```

**Comment by OhmSpectator** [comment]
> because there are still some deviations:

Do we have a good reason for these deviations?

**Comment by christoph-zededa** [comment]
> > because there are still some deviations:
> 
> Do we have a good reason for these deviations?

For bpftrace-compiler I just want to use the newer alpine as we will move to it anyways, so I didn't start with the old one.

For the other ones I don't know.

File: Makefile
Comment (OhmSpectator) [inline Makefile:467]: I don't think we should ignore other directories besides bpftracer. It's better to ask for these inconsistencies to be fixed.

File: Makefile
Comment (rene) [inline Makefile:468]: Why optee-os and bso-imx needs to be ignored?

File: Makefile
Comment (christoph-zededa) [reply Makefile:467]: and block this PR until then?

File: tools/dockerfile-from-checker/go.sum
Comment (OhmSpectator) [inline tools/dockerfile-from-checker/go.sum:1]: In Python, it would be fewer than 100 lines and no huge `go.mod` files. =(
But let it be if we already have it implemented...

File: tools/dockerfile-from-checker/main.go
Comment (rene) [inline tools/dockerfile-from-checker/main.go:99]: Would be good to have a minimal description of what the tool does...

File: Makefile
Comment (christoph-zededa) [reply Makefile:468]: tags differ for image lfedge/eve-cross-compilers:250abc77c8c39664905b66a2673102ec5cd3b056 in files /home/christoph/projects/eve/pkg/bsp-imx/Dockerfile and /home/christoph/projects/eve/pkg/pillar/Dockerfile

File: tools/dockerfile-from-checker/go.sum
Comment (christoph-zededa) [reply tools/dockerfile-from-checker/go.sum:1]: But can python evaluate the dockerfile variables?

File: Makefile
Comment (OhmSpectator) [reply Makefile:467]: You can just remove the ignoring and make it through errors every time so that the people will notice the issue =)

Comment (christoph-zededa) [reply Makefile:467]: But then yetus will fail on every PR.

File: Makefile
Comment (christoph-zededa) [reply Makefile:463]: Which other target should I use?
yetus is called by the GH actions.

File: Makefile
Comment (christoph-zededa) [reply Makefile:468]: tags differ for image lfedge/eve-cross-compilers:250abc77c8c39664905b66a2673102ec5cd3b056 in files /home/christoph/projects/eve/pkg/optee-os/Dockerfile and /home/christoph/projects/eve/pkg/pillar/Dockerfile

File: Makefile
Comment (OhmSpectator) [reply Makefile]: Argh... I didn't notice it's a part of Yetus target... Nope, I think it should be a dedicated target.

**Comment by OhmSpectator** [comment]
I would suggest to make it in two different PRs: the version bump (as the original version) and new check introduction.

File: Makefile
Comment (rene) [reply Makefile:473]:
```
pkg/bsp-imx/Dockerfile:23:FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:250abc77c8c39664905b66a2673102ec5cd3b056 AS cross-compilers
pkg/optee-os/Dockerfile:20:FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:250abc77c8c39664905b66a2673102ec5cd3b056 AS cross-compilers
pkg/pillar/Dockerfile:36:FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:2a1d062fce410865e7024a83de327a68e52db26c AS cross-compilers
```

They shouldn't differ, they should be the same as well, like eve container.... just bump them to the newest version and you can remove from the ignore list...

File: Makefile
Comment (rene) [reply Makefile:469]: A new dedicated target, then you can add it to GH actions. Also, the yetus GH action will not call `make yetus`, the execution is performed by the [yetus-test-patch-action](https://github.com/apache/yetus-test-patch-action), which calls the yetus test-patch directly from the source code tree....

**Comment by christoph-zededa** [comment]
> I would suggest to make it in two different PRs: the version bump (as the original version) and new check introduction.

@OhmSpectator and @rene What's the reason behind that besides that it creates more work for ME?

File: Makefile
Comment (christoph-zededa) [reply Makefile]: What about renaming `yetus` to `lint`? I don't want to go through all the hazzle of creating a new GH action ...

**Comment by deitch** [comment]
> I would suggest to make it in two different PRs: the version bump (as the original version) and new check introduction.

Same.

> What's the reason behind that besides that it creates more work for ME?

So they can be reviewed independently, easier to see in the future if something broke, easier to revert part if needed.

File: tools/dockerfile-from-checker/go.sum
Comment (OhmSpectator) [reply tools/dockerfile-from-checker/go.sum:1]: Ah, I got the idea... You parse AST there to get the values. Okay-okay.

**Comment by deitch** [comment]
I have a separate question, maybe I missed it in the comments. Why do we need to move dockerfile-from-checker from [eve-build-tools](https://github.com/lf-edge/eve-build-tools/tree/main/src/dockerfile-from-checker) to here? Isn't the point of that repo to be able to have tools that are useful, lightweight, may be used in multiple places, and can be updated without the whole heavy eve CI cycle? The "library" (or "modules") rationale?

**Comment by OhmSpectator** [comment]
> > I would suggest to make it in two different PRs: the version bump (as the original version) and new check introduction.
> 
> @OhmSpectator and @rene What's the reason behind that besides that it creates more work for ME?

We can have a bump of the hashes soon and properly implement the check with no hurry...

**Comment by deitch** [comment]
Separately, I am curious why we would bump eve-alpine in only 3 places, rather than all of them?

**Comment by OhmSpectator** [comment]
> I have a separate question, maybe I missed it in the comments. Why do we need to move dockerfile-from-checker from [eve-build-tools](https://github.com/lf-edge/eve-build-tools/tree/main/src/dockerfile-from-checker) to here? Isn't the point of that repo to be able to have tools that are useful, lightweight, may be used in multiple places, and can be updated without the whole heavy eve CI cycle? The "library" (or "modules") rationale?

By the way, I've just learned about this repo. Didn't pay attention to it before...
Now I need to understand the difference between 
* https://github.com/lf-edge/eve-build-tools/
* https://github.com/lf-edge/eve/tree/master/tools
* https://github.com/zededa/eve-tools

**Comment by deitch** [comment]
I can speak towards intent. How much those _really_ reflect reality is up for discussion:

* https://github.com/zededa/eve-tools - internal and private to Zededa 
* https://github.com/lf-edge/eve/tree/master/tools - tools that are tightly integrated into eve, and make no sense outside of eve
* https://github.com/lf-edge/eve-build-tools/ - small independent tools that may only be used in eve, but are generic enough to be elsewhere. Primarily, these are tools that don't need all of EVE CI to run, so make them in a dedicated repo that can be very quick and easy to update.

Does that help?

File: Makefile
Comment (OhmSpectator) [reply Makefile]: As @rene mentioned, the Yetus target is not called by any GH action. So, if we want to add it to GH actions, we have to add a new action anyway. Just don't do it if you don't want to. We can do it later.

**Comment by deitch** [comment]
#4340 should fix the `pkg/installer` package issues.

**Comment by deitch** [comment]
> I think the purpose of https://github.com/zededa/eve-tools and https://github.com/lf-edge/eve-build-tools/ is pretty much the same.

As long as stuff in eve-tools is appropriate for OSS, by all means.

> And I would say the tool for checking Docker hashes fits the EVE repo well.

I think it might be better in the separate repo. Unless it really is "eve special" and closely tied to EVE's build, it is better in eve-build-tools. Isn't it already there anyways?

**Comment by christoph-zededa** [comment]
> > What's the reason behind that besides that it creates more work for ME?
> 
> So they can be reviewed independently, easier to see in the future if something broke, easier to revert part if needed.

1. You already reviewed it here: https://github.com/lf-edge/eve-build-tools/pull/4
2. You can review it independently by going to this link: https://github.com/lf-edge/eve/pull/4334/commits/0f94dd8180425169e80b6a20c0450cc329bac9ae
3. `git revert` reverts a commit, not a PR.

**Comment by OhmSpectator** [comment]
> Isn't it already there anyways?

As @christoph-zededa mentioned before, it is. And we can reuse it from there in a new GH action later.

**Comment by deitch** [comment]
> git revert reverts a commit, not a PR

True, but PRs are part of a flow that people use.

**Comment by christoph-zededa** [comment]
> I have a separate question, maybe I missed it in the comments. Why do we need to move dockerfile-from-checker from [eve-build-tools](https://github.com/lf-edge/eve-build-tools/tree/main/src/dockerfile-from-checker) to here? 

Because I saw what point it is with the `dockerfile-add-scanner`, where you have to go to 3 entities to use a newer version:
* `eve-tools` repo
* docker hub
* the actual repo where you want to use a newer version

It is just too much overhead and hard to understand.

**Comment by christoph-zededa** [comment]
> > git revert reverts a commit, not a PR
> 
> True, but PRs are part of a flow that people use.

There is no flow to revert a PR.

File: Makefile
Comment (christoph-zededa) [reply Makefile]: Ah, I see. Can I add it back to the `make test` target?

Comment (OhmSpectator) [reply Makefile]: For local usage, sure. I would still like not to ignore the directories that do not have explicit requirements to be ignored. But it's also not an original scope of this PR and can be fixed later if you want.

Comment (christoph-zededa) [reply Makefile]: Thx, now I understand.

File: Makefile
Comment (christoph-zededa) [reply Makefile]: done

File: Makefile
Comment (christoph-zededa) [reply Makefile]: done: https://github.com/lf-edge/eve/actions/runs/11292579668/job/31408847842?pr=4334

File: Makefile
Comment (christoph-zededa) [reply Makefile]: done: https://github.com/lf-edge/eve/pull/4334#discussion_r1796910790

**Comment by christoph-zededa** [comment]
> > @OhmSpectator and @rene What's the reason behind that besides that it creates more work for ME?
> 
> We can have a bump of the hashes soon and properly implement the check with no hurry...

It is already there for nearly eleven months and nothing moved ...
(https://github.com/lf-edge/eve-build-tools/commits/main/src/dockerfile-from-checker )

File: Makefile
Comment (christoph-zededa) [reply Makefile]: @rene nope:
```
#18 [build-cross-target 2/3] RUN apk add --no-cache --allow-untrusted -X /packages build-base-"riscv64" build-base-armhf
Error: error building "lfedge/eve-optee-os:d031d1245591c210958b8cec260eaa804f05b157": error building for arch riscv64: failed to solve: process "/bin/sh -c apk add --no-cache --allow-untrusted -X /packages build-base-\"${EVE_TARGET_ARCH}\" build-base-armhf" did not complete successfully: exit code: 1
#18 0.043   build-base-armhf (no such package):
#18 0.043 ERROR: unable to select packages:
#18 0.046     required by: world[build-base-armhf]
#18 ERROR: process "/bin/sh -c apk add --no-cache --allow-untrusted -X /packages build-base-\"${EVE_TARGET_ARCH}\" build-base-armhf" did not complete successfully: exit code: 1
2024/10/11 13:02:40 error during command execution: error building "lfedge/eve-optee-os:d031d1245591c210958b8cec260eaa804f05b157": error building for arch riscv64: failed to solve: process "/bin/sh -c apk add --no-cache --allow-untrusted -X /packages build-base-\"${EVE_TARGET_ARCH}\" build-base-armhf" did not complete successfully: exit code: 1
```

**Comment by OhmSpectator** [comment]
> > > @OhmSpectator and @rene What's the reason behind that besides that it creates more work for ME?
> > 
> > 
> > We can have a bump of the hashes soon and properly implement the check with no hurry...
> 
> It is already there for nearly eleven months and nothing moved ... (https://github.com/lf-edge/eve-build-tools/commits/main/src/dockerfile-from-checker )

I didn't know about the tool (and about the entire repo))). Now I know and may find time to wrap into a GH Action.

**Comment by christoph-zededa** [comment]
> and may find time to wrap into a GH Action.

GH action is now there (and it was not as awful as I feared): https://github.com/lf-edge/eve/actions/runs/11292579668/job/31408847842?pr=4334

File: .github/workflows/lint.yml
Comment (OhmSpectator) [inline .github/workflows/lint.yml:4]: As for me, the name is too generic. I would be more specific and name it something like "Check Docker Hashes Consistency". So when it's red, we immediately see what exactly failed.

File: .github/workflows/lint.yml
Comment (OhmSpectator) [inline .github/workflows/lint.yml:29]: If we choose to run the check on all the files in the repo, not on the files affected by the PR for which the check runs, we should be able to ensure that all the hashes are consistent when we merge this PR (#4334). Are we?

File: .github/workflows/lint.yml
Comment (OhmSpectator) [inline .github/workflows/lint.yml:34]: Interesting. I'm just curious why this magic is necessary.

File: Makefile
Comment (OhmSpectator) [inline Makefile:463]: No way to avoid it?... Maybe let it fail now, and that way, force the owners to fix the inconsistency?...

File: .github/workflows/lint.yml
Comment (christoph-zededa) [reply .github/workflows/lint.yml:34]: Me neither ... (I got it from the go tests workflow)

File: Makefile
Comment (christoph-zededa) [reply Makefile:463]: If you really really really want a red pipeline on every PR, then let's introduce that in a new PR and not with this one, so that everybody gets warned publicly.

File: .github/workflows/lint.yml
Comment (christoph-zededa) [reply .github/workflows/lint.yml:4]: I thought we might add other linters in the future. Or do you want to introduce a new GH action for every new linter?

File: .github/workflows/lint.yml
Comment (OhmSpectator) [reply .github/workflows/lint.yml:29]: Yep, we run go tests for the entire repo, not for the files that have been changed. It's not a problem as we try to keep the go tests green.

File: .github/workflows/lint.yml
Comment (OhmSpectator) [reply .github/workflows/lint.yml:4]: I find it cleaner to have a named action per linter/check so the failed action is easier to understand.

File: .github/workflows/lint.yml
Comment (OhmSpectator) [reply .github/workflows/lint.yml:34]: @yash-zededa, @uncleDecart, do you have an idea of whether we need it here? )

File: Makefile
Comment (OhmSpectator) [reply Makefile:463]: Then, please, at least drop a TODO/FIXME comment close to these lines so we remember why this workaround was introduced.

File: .github/workflows/lint.yml
Comment (OhmSpectator) [reply .github/workflows/lint.yml:29]: Okay, I see the check is already green for this PR. Good. Document please the workaround for the current inconsistency, and that's enough (https://github.com/lf-edge/eve/pull/4334#discussion_r1797004499).

**Review by OhmSpectator — APPROVED** [review]
Run the tests

**Comment by OhmSpectator** [comment]
Why doesn't it run the tests >=(

**Review by yash-zededa — APPROVED** [review]

File: .github/workflows/lint.yml
Comment (christoph-zededa) [reply .github/workflows/lint.yml]: Renamed it.

File: Makefile
Comment (rene) [reply Makefile]: yup, you didn't bump to the latest one:

```diff
diff --git a/pkg/pillar/Dockerfile b/pkg/pillar/Dockerfile
index 17243cef9..cfc58c383 100644
--- a/pkg/pillar/Dockerfile
+++ b/pkg/pillar/Dockerfile
@@ -33,7 +33,7 @@ ARG BUILD_PKGS_BASE
 RUN BUILD_PKGS="${BUILD_PKGS_BASE}" eve-alpine-deploy.sh
 
 # hadolint ignore=DL3029
-FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:2a1d062fce410865e7024a83de327a68e52db26c AS cross-compilers
+FROM --platform=${BUILDPLATFORM} lfedge/eve-cross-compilers:250abc77c8c39664905b66a2673102ec5cd3b056 AS cross-compilers
 
 # will use several packages from target arch and copy them to sysroot
 # hadolint ignore=DL3006
```

and then all eve-cross-compilers should stay with the same hash.....

File: Makefile
Comment (rene) [reply Makefile:464]: @OhmSpectator , @christoph-zededa , please, see https://github.com/lf-edge/eve/pull/4334#discussion_r1797043925

**Comment by OhmSpectator** [comment]
Now I really like it.
But I'm a little bit confused. How is this PR affected by https://github.com/lf-edge/eve/pull/4340? Shouldn't we merge this one before #4340?

**Comment by christoph-zededa** [comment]
> Now I really like it. But I'm a little bit confused. How is this PR affected by #4340? Shouldn't we merge this one before #4340?

Once #4340 is merged, the `installer/Dockerfile` can be removed from the list of ignored Dockerfiles. Then only bpftrace is left (until we bump everything to alpine 3.20).

**Comment by deitch** [comment]
Probably 4340 first, but I can rebase that one next week if this goes in.

**Comment by OhmSpectator** [comment]
> Probably 4340 first, but I can rebase that one next week if this goes in.

Let's merge this PR first, as it's already in good shape. And rebase #4334 later (it should also remove the ignoring of the `pkg/installer`).

**Comment by OhmSpectator** [comment]
Why are the Eden tests skipped all the time?.. @yash-zededa, any idea?

**Review by OhmSpectator — APPROVED** [review]

**Comment by yash-zededa** [comment]
> Why are the Eden tests skipped all the time?.. @yash-zededa, any idea?

It's because of the WF modifications. 

@christoph-zededa would it be possible to create a separate PR for the WF and Makefile

**Comment by OhmSpectator** [comment]
> > Why are the Eden tests skipped all the time?.. @yash-zededa, any idea?
> 
> It's because of the WF modifications.
> 
> @christoph-zededa would it be possible to create a separate PR for the WF and Makefile

@yash-zededa, do we skip the tests run if workflow files are changed? I would expect the tests to be skipped if ONLY workflows are changed... Do you think we could fix it?

**Comment by yash-zededa** [comment]
> > > Why are the Eden tests skipped all the time?.. @yash-zededa, any idea?
> > 
> > 
> > It's because of the WF modifications.
> > @christoph-zededa would it be possible to create a separate PR for the WF and Makefile
> 
> @yash-zededa, do we skip the tests run if workflow files are changed? I would expect the tests to be skipped if ONLY workflows are changed... Do you think we could fix it?

I think so, I will work on the logic where in we run tests for PR which has WF as well as code changes but ideally for all the new WF’s/modification to the existing one’s should be part of separate PR since it will help us with a clean history.

**Comment by christoph-zededa** [comment]
> 
> @christoph-zededa would it be possible to create a separate PR for the WF and Makefile

like https://github.com/lf-edge/eve/pull/4341 ?

**Comment by OhmSpectator** [comment]
> > @christoph-zededa would it be possible to create a separate PR for the WF and Makefile
> 
> like #4341 ?

Merged. Rebase this one, please.

**Review by OhmSpectator — APPROVED** [review]

File: Makefile
Comment (OhmSpectator) [inline Makefile:464]: Should it be rebased to use the new variables introduced by @deitch, I guess?

Comment (deitch) [reply Makefile:464]: Yup. And I just opened a PR based on that to bump the alpines as well, but _remove_ `pkg/installer` from the exceptions list 😄 

Not sure if that helps, hinders or is neutral to this one. Feel free to throw mine away if needed.

Comment (christoph-zededa) [reply Makefile:464]: I think it's best to wait for rebase until https://github.com/lf-edge/eve/pull/4348 is merged ...

Comment (OhmSpectator) [reply Makefile]: > I think it's best to wait for rebase until #4348 is merged ...

And that PR is dependent on https://github.com/lf-edge/eve/pull/4350

**Comment by OhmSpectator** [comment]
@christoph-zededa, it's time to rebase! =) #4348 is merged!

**Review by eriknordmark — APPROVED** [review]
LGTM

**Pull request merged**
